[KB] Update: Revise required workflows to repository rulesets (#1226)
returntocorp/semgrep-docs: docs/kb/semgrep-ci/github-repository-rulesets-semgrep.md

---
tags:
  - GitHub
  - Repository rulesets

> description: Setting up GitHub required workflows for Semgrep scans.

Reviewer: ```suggestion
description: Set up GitHub required workflows to efficiently implement Semgrep scans across many repositories.
```
I haven't been doing a good job catching this, but we use simple present for the description. 

"

(it makes sense when you see the kb cards with the descriptions)

---

# Using GitHub repository rulesets to implement Semgrep

> If your GitHub organization has hundreds or thousands of repositories, setting up Semgrep by adding the `.github/workflows/semgrep.yml` file to all repositories can be a time-consuming task. You can instead use [repository rulesets](https://docs.github.com/en/enterprise-cloud@latest/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/creating-rulesets-for-a-repository#introduction) to add Semgrep as a workflow that is [required to pass before merging](https://docs.github.com/en/enterprise-cloud@latest/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/available-rules-for-rulesets#require-workflows-to-pass-before-merging). 

Reviewer: ```suggestion
Use [GitHub repository rulesets](https://docs.github.com/en/enterprise-cloud@latest/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/creating-rulesets-for-a-repository#introduction) to quickly implement Semgrep scans across hundreds or thousands of repositories in your GitHub organization. Repository rulesets enable you to add a Semgrep scan to many repositories as a workflow that is [required to pass before merging](https://docs.github.com/en/enterprise-cloud@latest/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/available-rules-for-rulesets#require-workflows-to-pass-before-merging).
```

I feel like there's two things going on here in this paragraph: 
1. Repository rulesets make it easy to add Semgrep scans
2. Repository rulesets require that the workflow must pass. 

We may want to separate (2) as its own paragraph. 

Also, I think there are two things we should mention early -  mainly that "repository rulesets make it easy to add Semgrep scans to PRs by having 1 repository rulesets JSON file exported to many repos" (my understanding is based on the helpful documentation link you provided).

> 

Reviewer: ```suggestions
Repository rulesets **reference** a workflow file. A workflow file defines the processes of a GitHub Actions job. This article uses a basic `semgrep.yml` workflow file to run Semgrep scans in GitHub Actions.

The repository ruleset can **target** many repositories at once. By doing this, you are able to reference one workflow file for many repositories.

> This allows you to set up a centralized repository with a single Semgrep workflow file that can be applied to as many repositories as desired. Formerly, this feature was called [required workflows](https://github.blog/changelog/2023-08-02-github-actions-required-workflows-will-move-to-repository-rules/).

Reviewer: The workflow file can be placed in a centralized repository. Formerly, this feature was called [required workflows](https://github.blog/changelog/2023-08-02-github-actions-required-workflows-will-move-to-repository-rules/).

## Setting up the central Semgrep scan workflow

To use the Semgrep workflow in other repositories, you can either create a new repository with the Semgrep workflow file, or add it to an existing repository where you store common workflows. 

> When a workflow is required to pass before merging, the workflow file needs to be in a repository that matches the widest visibility of the repositories you want to run it in. For example, if you want to run Semgrep on public, internal, and private repositories, the repository containing the workflow file must be public.

Reviewer: Just curious - what if the workflow is not required to pass?

Author: The only option for repository rulesets & workflows is to require them to pass before merging. If you want to do something else you'd have to use some other mechanism.

Reviewer: In that case I think we can say "the workflow fiiles neds to be run..." without the prior clause, because there's no other case.

8. Then, select the branch, tag, or commit to use.
9. In the **Pick a workflow file** field, click and select the Semgrep workflow you created in [Setting up the central Semgrep scan workflow](#setting-up-the-central-semgrep-scan-workflow).
10. Click **Add workflow**.
    ![Require workflows to pass before merging](/img/kb/semgrep-workflow-require-pass.png)
11. Click **Create** to create the ruleset.

You can refer to GitHub's [Creating rulesets for repositories in your organization](https://docs.github.com/en/enterprise-cloud@latest/organizations/managing-organization-settings/creating-rulesets-for-repositories-in-your-organization) for more general guidance on creating a ruleset for your organization.

## Verify by creating a pull request

After setup, hen you create a pull request in a repository where this ruleset applies, the Semgrep workflow runs for the PR even if there is no `.github/workflows/semgrep.yml` file in that repository.

The required workflow allows merge if the scan is successful, or blocks the pull request if the scan has blocking findings.

![Example PR with successful required workflow](/img/kb/semgrep-workflow-pr-example.png)

## Limitations

> Workflows required by repository rulesets are only triggered by `pull_request` or `merge_group` events. When triggered for a pull request, Semgrep runs a [diff-aware scan] and only scans changed files. Therefore, to run a full scan for your organization's repositories (scan all files), you would need to choose a different approach, such as [reusable workflows](/docs/kb/semgrep-ci/github-reusable-workflows-semgrep).

Reviewer: Are we, in effect, recommending that users use both reusable workflows AND repository rulesets to have full coverage (both full scans and diff scans)? Put another way, what is the ideal path for the user to take when implementing Semgrep scans for 100s or 1000s of repositories?

Reviewer: (Also, I think the [diff-aware] scan is meant to be a link?

Author: yes totally was, ty for the reminder
